UART for POSIX_ARCH

aescolar opened the pull request:
Added a new UART driver for posix arch boards. The driver can be configured to either attach to a new
pseudo-terminal, or to connect to the process stdin-out (only meant for automatic testing, not for interactive use)

--------------------
Fixes #10190 
Fixes #10127 by working around #10191

aescolar committed:
uart: Add driver for posix arch

Added a new UART driver for posix arch boards.
The driver can be configured to either attach to a new
pseudo-terminal, or to connect to the invoking shell
stdin-out.

Signed-off-by: Alberto Escolar Piedras <alpi@oticon.com>

aescolar committed:
subsys: shell: support posix arch

The posix arch does not compile either of Zephyr's libc,
so _prf() is not avaliable

Signed-off-by: Alberto Escolar Piedras <alpi@oticon.com>

aescolar committed:
subsys: shell: Add missing dependency to SERIAL

The shell subsystem, as it is today, depends on having a UART,
therefore let's add the dependency explicitly in its Kconfig

Fixes #10190

Signed-off-by: Alberto Escolar Piedras <alpi@oticon.com>

aescolar commented:
One clarification to a question you will probably have:
The default is, purposely, to attach the UART to its own pty.
The option to map it to STDIN/OUT is really not the intended use-case. 

The reason for this is that the shell assumes it has access to a terminal thru a **raw** driver, without any of the standard terminal niceties. 

That means that if I we wanted to provide some reasonable setup thru stdin/out for interactive use, we would need to reconfigure the terminal into a non-sane mode (no echo, no line input, no Ctrl+X raising signals, etc..).
That can be done easily, but then we have 2 issues:
* We need to be sure to leave the terminal as it was on SIGSTP (Ctrl+Z) and bring it back to funny mode in SIGCONT. Otherwise the terminal is not usable if/when people leaves the native_posix executable in the background. This is double with a bit of extra code.
* We need to be sure to leave the terminal back how it was on exit. And here is the big problem. When things end gracefully enough, we can bring the terminal back. But given that native_posix is meant as a debugging/development aid, it is likely/normal that things will crash horribly (e.g. segfault). If we don't bring the terminal back it would be left in an unusable and unrecoverable state for 99.9% of the users.

aescolar commented:
You can for example try the shell sample app as:
```
cmake ../samples/subsys/shell/shell_module/  -GNinja -DBOARD=native_posix  -DCONFIG_NO_OPTIMIZATIONS=y
ninja
zephyr/zephyr.exe --attach_uart #assuming you have xterm and screen installed
```
If you run any of the commands which require options without an option you will trigger #10195

aescolar commented:
recheck

ulfalizer commented on subsys/shell/Kconfig:
```diff
@@ -38,6 +38,7 @@ menuconfig SHELL
 	bool "Enable shell"
 	select LOG_RUNTIME_FILTERING
 	select POLL
+	select SERIAL
```
`depends on` might be better here. `select`ing symbols with prompts or dependencies often leads to two common issues:

 - `select` ignores dependencies. If someone adds (unsatisfied) dependencies to `SERIAL`, this will still force it on (but at least you get a warning).

 - With `select`, you can't immediately tell when looking at `SERIAL` that there's other ways for it to get enabled, or that `CONFIG_SHELL=y` also enables `SERIAL`. This can lead to twisty and hard-to-follow Kconfig files if there's a lot of `select`ing going on.

Usually, it's best to limit `select` to simple invisible "helper" symbols (`select CPU_SUPPORTS_FOO` and the like). It's great for that.

I whitelisted the select-without-satisfied-dependencies warning when I was turning warnings into errors, because there was a lot of it going on, and a lot of it seemed to be from `CONSOLE`-related symbols. Should probably refactor a bunch of those to use `depends on` instead. :)

ulfalizer replied:
This sentiment also appears in `kconfig-language.txt` btw:

```
select should be used with care. select will force a symbol to a value without
visiting the dependencies. By abusing select you are able to select a symbol
FOO even if FOO depends on BAR that is not set. In general use select only for
non-visible symbols (no prompts anywhere) and for symbols with no dependencies.
That will limit the usefulness but on the other hand avoid the illegal
configurations all over.
```

Common sense overrides it though, but in practice, `select` very often leads to those issues.

aescolar replied:
Agreed.

dbkinder commented on boards/posix/native_posix/doc/board.rst:
```diff
@@ -19,6 +19,8 @@ you use native host tools for compiling, debugging, and analyzing your
 Zephyr application, eliminating the need for architecture-specific
 target hardware in the early phases of development.
 
+This board provides a few peripherals. See `Peripherals`_ for more information.
```
Some examples would be nice here:

    This board provides a few peripherals such as an Ethernet driver and UART.  See ...

dbkinder commented on boards/posix/native_posix/doc/board.rst:
```diff
@@ -458,8 +460,8 @@ The following peripherals are currently provided with this board:
   - Feed any input from the native application ``stdin`` to a possible
-    running :ref:`Shell`. For more information refer to the section
-    `Shell support`_.
+    running legacy shell. For more information refer to the section
```
Not the new shell, eh?  Will the new and legacy shells work together or are they mutually exclusive?

aescolar replied:
> Not the new shell, eh? 

Not with that console driver. The new shell is not using ``CONSOLE`` , but hooks directly into the UART driver. So all the ``CONSOLE`` backends we had are of no use with the new shell.

> Will the new and legacy shells work together or are they mutually exclusive?

I guess the question should be for the new shell developers. But my understanding is that, today:
* The modules using the legacy shell are being ported to use the new shell. So the legacy shell will be purposeless quite soon (unless some odd app uses it).
* The new shell does not know or care about the old shell, and does not intend to avoid any conflict with it. If the old shell was driven thru UART they will certainly conflict if compiled together.
* In case of native_posix, I think you may still be able to compile the old shell in parallel with the new one by using this native console driver for the old shell, and the new UART driver for the new shell.

jarz-nordic replied:
AFAI legacy shell will be removed completely

dbkinder commented on boards/posix/native_posix/doc/board.rst:
```diff
@@ -538,11 +540,40 @@ The following peripherals are currently provided with this board:
 
-When the :ref:`Shell` subsystem is compiled with your application, the native
-standard input (``stdin``) will be redirected to the shell.
+When the legacy :ref:`Shell` subsystem is compiled with your application,
```
``:ref:`shell` `` will be a reference to the **new** shell documentation and not to the **legacy** shell subsystem.  The shell documentation is changing with PR #10216, so your PR's references may break when 10216 is merged (or not work until it is). This should probably change to:

    When the :ref:`legacy shell` subsystem is compiled ...

aescolar replied:
Yep.. I saw that PR coming after I sent mine. So I guess my options are 
a. to leave the old link which is correct today, and if that PR is merged before mine, correct mine
b. remove all links to ``Shell`` all together
c. Point to the new shell doc and stall this PR.. => don't like this one.

aescolar committed:
tests: uart: add extra config for native_posix

When running a UART test in CI, it is better to configure
the nativeposix UART to be mapped to the process STDIN/OUT.

Signed-off-by: Alberto Escolar Piedras <alpi@oticon.com>

aescolar commented:
@jarz-nordic: are you ok with the fixes in the shell files?

aescolar commented:
@nashif , @jukkar , @jhedberg : Please take a quick look at the code, and tell if it there is something obviously horrible or not. This code is adding a new optional feature meant for testing, so I do not think we need an extremely thorough review. (I do not expect anybody to review the particular terminal driver flags combination...). 

If any of you can think of a better reviewer for this, just add him.

jukkar reviewed (approved):
LGTM

jarz-nordic commented:
@aescolar : sorry for late review, I had a busy day.

pfalcon commented on subsys/shell/shell_fprintf.c:
```diff
@@ -32,7 +32,7 @@ static int out_func(int c, void *ctx)
 void shell_fprintf_fmt(const struct shell_fprintf *sh_fprintf,
 		       const char *fmt, va_list args)
 {
-#ifndef CONFIG_NEWLIB_LIBC
```
Commit message for this change:

> The posix arch does not compile neither of Zephyr's libc,

sounds a bit strange in the standard English, should be "The posix arch does not compile *either* of Zephyr's libc's,"

pfalcon commented:
Is this supposed to fix https://github.com/zephyrproject-rtos/zephyr/issues/5941 ? If no, why? If yes, doc clauses like:

> :option:`CONFIG_NATIVE_UART_0_ON_OWN_PTY` or
:option:`CONFIG_NATIVE_UART_0_ON_STDINOUT`
Note that for interactive use, the first option should be chosen. The
second option is only intended for automated testing or feeding/piping other
processes output to the UART.

Seem strange/limiting. I wanted to check myself, but hit another obstacle: https://github.com/zephyrproject-rtos/zephyr/issues/10243

pfalcon commented:
> Is this supposed to fix #5941 ?

Well, it doesn't. There's no irq mode support, only busy-polling.

aescolar commented:
> Seem strange/limiting

It just means that for interactive use, you have a dedicated TTY for the UART, instead of taking over the one from where you launch the zephyr app.

pfalcon commented:
> It just means that for interactive use, you have a dedicated TTY for the UART, instead of taking over the one from where you launch the zephyr app.

In a typical embedded situation, there's just one UART for user to connect to and do everything (provide input and get output, see logging, etc.). I'd expect the same functionality to be available on an emulated Zephyr port (at least available, if not default).

aescolar commented:
I think you missunderstood me. When you configure the native_poxis UART with CONFIG_NATIVE_UART_0_ON_OWN_PTY (default), you get a new linux /dev/pts mapped to the Zephyr UART. If you use the --atach_uart command, you even get a new window with a terminal emulator connected to it pop automagically.

pfalcon commented:
Right, and if CONFIG_NATIVE_UART_0_ON_STDINOUT is defined, your docs say that it "is only intended for automated testing or feeding/piping other
processes output to the UART." I don't see why it would be intended only for that. Anyway, that can be handled later.

aescolar commented:
> I don't see why it would be ...

If you try to use the shell interactively in that way, you will find yourself with a lot of annoyances, like both the Linux terminal driver and Zephyr shell echoing, the terminal capturing the Ctrl+x combinations (which the shell expects to get raw), the terminal replacing the \r with a \n, while the shell only reacts to \r as end of line, and so forth. That's why I warn people in advance that it is not going to be pleasant, and guide them to the other option which provides what I think most people will expect.

pfalcon commented:
> If you try to use the shell interactively in that way, you will find yourself with a lot of annoyances, like both the Linux terminal driver and Zephyr shell echoing

Ack, I see. As I hinted above, I quickly tried this patch with samples/subsys/console/echo/, but that failed as it required interrupt-driven UART, and I didn't look further so far. Now the situation is clear, thanks. I'd be keen to do something about it, but will look into it later.

ulfalizer reviewed (approved):
`select` nit fixed.

aescolar commented:
@dbkinder : Is your review stale? or do you want to hold merging this one until the shell doc PR is fixed and merged, and this one would point to the new pages?

jukkar commented:
I would like to see this merged asap, the net-shell conversion at #9825 is currently blocked and needs this fix.

aescolar commented:
This last push is just a commit message fix to address the neither->either issue. No other change.

dbkinder reviewed (approved):
I'll get out of the way for this PR and we'll deal with the doc changes after the new shell code and docs are merged.

aescolar commented:
@nashif / @carlescufi : If either of you does not have anything against, feel free to press the button.
